File: content/problems/1770-maximum-score-from-performing-multiplication-operations/solutions/operation-left-count-dp/Solution.java
class Solution {
    private int[] nums,multipliers;
    private Integer[][] memo;
    private int dfs(int step,int left){
        if(step==multipliers.length)return 0;
        if(memo[step][left]!=null)return memo[step][left];
        int right=nums.length-1-(step-left);
        int takeLeft=nums[left]*multipliers[step]+dfs(step+1,left+1);
        int takeRight=nums[right]*multipliers[step]+dfs(step+1,left);
        return memo[step][left]=Math.max(takeLeft,takeRight);
    }
    public int maximumScore(int[] nums,int[] multipliers){
        this.nums=nums;
        this.multipliers=multipliers;
        memo=new Integer[multipliers.length][multipliers.length];
        return dfs(0,0);
    }
}

File: content/problems/1770-maximum-score-from-performing-multiplication-operations/solutions/operation-left-count-dp/solution.cpp
class Solution {
    vector<int> nums,multipliers;
    vector<vector<int>> memo;
    int dfs(int step,int left){
        if(step==(int)multipliers.size())return 0;
        int& answer=memo[step][left];
        if(answer!=INT_MIN)return answer;
        int right=nums.size()-1-(step-left);
        return answer=max(nums[left]*multipliers[step]+dfs(step+1,left+1),nums[right]*multipliers[step]+dfs(step+1,left));
    }
    public:int maximumScore(vector<int>& input,vector<int>& factors){
        nums=input;
        multipliers=factors;
        memo.assign(factors.size(),vector<int>(factors.size(),INT_MIN));
        return dfs(0,0);
    }
};
